internal/cmd/update_test.go

```diff
@@ -0,0 +1,542 @@
+	"strings"
+	"testing"
+
+	"cdr.dev/coder-cli/pkg/clog"
+	"cdr.dev/slog/sloggers/slogtest/assert"
+	"github.com/manifoldco/promptui"
+	"github.com/spf13/afero"
+	"golang.org/x/xerrors"
+)
+
+const (
+	fakeExePathLinux     = "/home/user/bin/coder"
+	fakeExePathWindows   = `C:\Users\user\bin\coder.exe`
+	fakeCoderURL         = "https://my.cdr.dev"
+	fakeNewVersion       = "1.23.4-rc.5+678-gabcdef-12345678"
+	fakeOldVersion       = "1.22.4-rc.5+678-gabcdef-12345678"
+	filenameLinux        = "coder-cli-linux-amd64.tar.gz"
+	filenameWindows      = "coder-cli-windows.zip"
+	fakeGithubReleaseURL = "https://api.github.com/repos/cdr/coder-cli/releases/tags/v1.23.4-rc.5"
+)
+
+var (
+	apiPrivateVersionURL  = fakeCoderURL + apiPrivateVersion
+	fakeError             = xerrors.New("fake error for testing")
+	fakeNewVersionJSON    = fmt.Sprintf(`{"version":%q}`, fakeNewVersion)
+	fakeOldVersionJSON    = fmt.Sprintf(`{"version":%q}`, fakeOldVersion)
+	fakeAssetURLLinux     = "https://github.com/cdr/coder-cli/releases/download/v1.23.4-rc.5/" + filenameLinux
+	fakeAssetURLWindows   = "https://github.com/cdr/coder-cli/releases/download/v1.23.4-rc.5/" + filenameWindows
+	fakeGithubReleaseJSON = fmt.Sprintf(`{"assets":[{"name":%q,"browser_download_url":%q},{"name":%q,"browser_download_url":%q}]}`, filenameLinux, fakeAssetURLLinux, filenameWindows, fakeAssetURLWindows)
+)
+
+func Test_updater_run(t *testing.T) {
+	t.Parallel()
+
+	//  params holds parameters for each test case
+	type params struct {
+		ConfirmF       func(string) (string, error)
+		Ctx            context.Context
+		Execer         *fakeExecer
+		ExecutablePath string
+		Fakefs         afero.Fs
+		HTTPClient     *fakeGetter
+		OsF            func() string
+		VersionF       func() string
+	}
+
+	// fromParams creates a new updater from params
+	fromParams := func(p *params) *updater {
+		return &updater{
+			confirmF:       p.ConfirmF,
+			execF:          p.Execer.ExecF,
+			executablePath: p.ExecutablePath,
+			fs:             p.Fakefs,
+			httpClient:     p.HTTPClient,
+			osF:            p.OsF,
+			versionF:       p.VersionF,
+		}
+	}
+
+	run := func(t *testing.T, name string, fn func(t *testing.T, p *params)) {
+		t.Logf("running %s", name)
+		ctx := context.Background()
+		fakefs := afero.NewMemMapFs()
+		execer := newFakeExecer(t)
+		execer.M["brew --prefix"] = fakeExecerResult{[]byte{}, os.ErrNotExist}
+		params := &params{
+			// This must be overridden inside run()
+			ConfirmF: func(string) (string, error) {
+				t.Errorf("unhandled ConfirmF")
+				t.FailNow()
+				return "", nil
+			},
+			Execer:         execer,
+			Ctx:            ctx,
+			ExecutablePath: fakeExePathLinux,
+			Fakefs:         fakefs,
+			HTTPClient:     newFakeGetter(t),
+			// Default to GOOS=linux
+			OsF: func() string { return goosLinux },
+			// This must be overridden inside run()
+			VersionF: func() string {
+				t.Errorf("unhandled VersionF")
+				t.FailNow()
+				return ""
+			},
+		}
+
+		fn(t, params)
```

Comment:
```suggestion
		t.Run(name, func(t *testing.T) {
			ctx := context.Background()
			fakefs := afero.NewMemMapFs()
			execer := newFakeExecer(t)
			execer.M["brew --prefix"] = fakeExecerResult{[]byte{}, os.ErrNotExist}
			params := &params{
				// This must be overridden inside run()
				ConfirmF: func(string) (string, error) {
					t.Errorf("unhandled ConfirmF")
					t.FailNow()
					return "", nil
				},
				Execer:         execer,
				Ctx:            ctx,
				ExecutablePath: fakeExePathLinux,
				Fakefs:         fakefs,
				HTTPClient:     newFakeGetter(t),
				// Default to GOOS=linux
				OsF: func() string { return goosLinux },
				// This must be overridden inside run()
				VersionF: func() string {
					t.Errorf("unhandled VersionF")
					t.FailNow()
					return ""
				},
			}

			fn(t, params)
		})
```

and possibly put `t.Parallel()` first inside t.Run if these tests can be run in parallel

Reply:
this diff is awful, tldr I wrapped everything in t.Run